dom/events/scrolling/wheel-event-groups.html

```diff
@@ -0,0 +1,134 @@
+  width: 100%;
+  height: 4000px;
+  background: red;
+}
+
+</style>
+<head>
+<body onload=runTest()>
+<div id="firstRootSpacer" class="spacer" style="background: cyan"></div>
+<div id="secondRootSpacer" class="spacer" style="background: magenta"></div>
+<div id="scrollableDiv">
+  <div id="innerDiv">
+    <div id="firstInnerSpacer" class="spacer" style="background: green">
+    </div>
+    <div id="secondInnerSpacer" class="spacer" style="background: blue">
+    </div>
+  </div>
+</div>
+</body>
+
+<script>
+
+let wheelEventTargets = [];
+window.addEventListener("wheel", e => {
+  wheelEventTargets.push(e.target)
+}, {passive: true});
+
+function wheelEventTargetsEqual(expectedTarget, errorString) {
+  wheelEventTargets.forEach((wheelEventTarget) => {
+    assert_equals(wheelEventTarget, expectedTarget, errorString);
+  });
+}
+
+let tests = [
+  {
+    origin: "viewport",
+    scrollX: 40,
+    scrollY: 2,
+    scrollingElement: document.scrollingElement,
+    targetElement: firstRootSpacer,
+    title: 'Wheel events are grouped and are not sent to a child element',
+  },
+  {
+    origin: "viewport",
+    scrollX: 40,
+    scrollY: 27,
+    scrollingElement: document.scrollingElement,
+    targetElement: secondRootSpacer,
+    title: 'Wheel events are grouped and are not sent to a child scroll frames',
+  },
+  {
+    origin: innerDiv,
+    scrollX: 40,
+    scrollY: 2,
+    scrollingElement: scrollableDiv,
+    targetElement: innerDiv,
+    title: 'Wheel events are not bound to a scroll frame',
+  },
+  {
+    origin: secondInnerSpacer,
+    scrollX: 40,
+    scrollY: 2,
+    scrollingElement: scrollableDiv,
+    targetElement: secondInnerSpacer,
+    title: 'Wheel event groups beyond the origin bound have the same target',
+  },
+];
+
+function runTest() {
+  tests.forEach((testInfo) => {
+    promise_test(async (t) => {
+      await waitForCompositorCommit();
+
+      // Wait some extra time to ensure nothing from the prior test impacts this one.
+      await new Promise(resolve => setTimeout(resolve, 1000));
+
+      wheelEventTargets = [];
+
+      // Scroll past the boundary of the original element to ensure all events in
+      // group have the same target.
+      //await new test_driver.Actions()
+      await new test_driver.Actions()
+        .addWheel("wheel1")
+        .scroll(testInfo.scrollX, testInfo.scrollY, 0, 30, {origin: testInfo.origin})
+        .pause(1)
+        .scroll(testInfo.scrollX, testInfo.scrollY, 0, 30, {origin: testInfo.origin})
+        .pause(1)
+        .scroll(testInfo.scrollX, testInfo.scrollY, 0, 30, {origin: testInfo.origin})
+        .send();
+
+      await waitForAnimationEnd(() => { return testInfo.scrollingElement.scrollTop; });
+      await waitForCompositorCommit();
+
+      testInfo.scrollingElement.scrollTo({top: 0, left: 0, behavior: "instant"});
+
+      wheelEventTargetsEqual(testInfo.targetElement,
+                             "All wheel events in the group have the same element");
```

Comment:
> Could you rewrite these tests to collect Event.target of wheel events with window.addEventListener into an array and compare each array item with what you expect?

Went about doing this in a slightly different manner. All the wheel event targets for a given test should be in the same wheel event group, so they all should be the same. As a result, we don't need to check `wheelEventTargets[0] == expectedTarget`, we can just iterate through all the wheel event targets we collected in the testcase and ensure they're the expected target.